### ISSUE TITLE
404 Error

### ISSUE THREAD

**Issue by taylorbourne** [frozeman/meteor-build-client#18]
I'm running the following command to run a build

```
meteor-build-client ../build -p ""
```

When I check the console after uploading to my webserver I get the following:

```
Failed to load resource: the server responded with a status of 404 (Not Found) 
http://urltomysite.com/sockjs/info?cb=m69amdvokn
```

**Comment by frozeman** [comment]
Thats normal, as meteor wants to connect to its server, but can't. We can't avoid this. But you shouldn't be seeing more than one of this errors, as we are calling `Meteor.disconnect()` in the index.html

**Comment by taylorbourne** [comment]
Ah ok awesome.. Yeah I'm only seeing it call and fail one time. I am seeing an issue where the JavaScript hangs up on initial page view. For example, if I have someone hit the site for the first time the page hangs. When they hit refresh everything loads as normal. I'll investigate further to see if the issues are related, if not I'll open a new issue. Thanks for the quick reply!

**Comment by GeoffreyBooth** [comment]
I solved this issue by creating a file `public/sockjs/info` with the contents:

```json
{"websocket": false}
```

With _any_ file in `public/sockjs/info`, there’s no 404; but Meteor will choke unless it’s a valid JSON file. The beginning of the normal Meteor `sockjs/info` is `{"websocket": true`, so I figured `{"websocket": false}` would be a good replacement.

Perhaps this should be part of the build process? Either as the default, or an option?

**Comment by frozeman** [comment]
Interesting solution, but i font really follow.
So you want to put this file in the `public/sockjs/info` folder. But what if you use your meteor build client without a server at all? So simple run the `index.html`?

Apart from that, it sounds like a good idea

So what is the file name supposed to be?

**Comment by GeoffreyBooth** [comment]
`info` is a file, not a folder. Just a file with no extension, inside `public/sockjs/`.

When Meteor build client makes the build, the contents of `public` are copied up to the root level, so you end up with a build folder like `index.html`, `blah256.js`, `blah256.css`, `sockjs/info`.

My use case is for hosting on Apache, but as far as I know this should work the same on Nginx.

**Comment by frozeman** [comment]
Hm, there could be the case that this prevents normal connection, if you actually would like o connect back too the serving server.

@duffar could you test this, if it interferes with your use case?

**Comment by jeffryan** [comment]
Sorry for the slow response here.

If your thinking of putting this into the build tool, it should only be done if you don't plan to connect to any remote backend. This is basically @frozeman 's scenario.

If you plan to eventually connect to a backend server (like we do) I think you should leave it up to the application or the backed server to specify the sockjs info contents. The server is basically telling the meteor front end what kind of sockjs connections will be supported/allowed and then the front end will know what kind of connect strategies to use.

By putting the sockjs file into the public folder you are essentially pre-loading the answer into the meteor client and it won't even go out to the remote server to look for it. Definitely wouldn't work for our scenario.

**Comment by jeffryan** [comment]
I'm afraid I need to retract my statement that it would break our scenario... I added a public/sockjs/info file to my meteor project with the contents:  
    {"websocket": false}   
and then rebuilt the app and the front end still seems to go back to our server and requests the /sockjs/info file. I must say I'm perplexed as to why... especially if thats what @GeoffreyBooth did to prevent meteor from requesting the file. I'll try some more to identify why.

**Comment by GeoffreyBooth** [comment]
Adding the `public/sockjs/info` file doesn’t prevent your app from requesting that path; it just _prevents a 404 error,_ since there is a file at that path for the server to return. And since the file is valid JSON, Meteor can parse it and therefore Meteor doesn’t throw some different error.

I think if we’re hard-coding `Meteor.disconnect()` within a `<script>` block in the `<head>`, then by definition we’re assuming that there’s no Meteor server to connect to via websockets. I think the default build behavior should be to include that `Meteor.disconnect()` as well as `sockjs/info` with `{"websocket": false}`, _but_ there should be an option for leaving out `disconnect` and `sockjs/info`. That way if someone wants to use Meteor Build Client somewhere where the _is_ a Meteor app server-side, they can without issues; but I think most people using Meteor Build Client probably have either no server-side code or it’s something else like PHP, and we want to turn off Meteor’s default websockets.

**Comment by jeffryan** [comment]
>Adding the public/sockjs/info file doesn’t prevent your app from requesting that path; it just prevents a 404 error, since there is a file at that path for *the server* to return.

Thanks for the explanation but I guess I am still a bit confused. If the info file is provided by *the server* how does Meteor Build Client play a part in that since it is only building the client portion of the application? Or are you saying it is packaged into the client as a static part and used as a fallback when the server gives a 404. Or something else entirely ;)

You likely understand this processing better than me so please forgive my ignorance.

I agree with you on why most people use Meteor Build Client: 
  1. No backend app at all
  2. Alternate backend app in php etc

We fall into the second camp in that we are building a reactive meteor front end that is served from an embedded device - which doesn't support the meteor server but is generally programmable. By including a web server and supporting DDP over websockets with a little sockjs crud we're able to make it work. I don't know if we're unique in this way or not but I suspect as meteor popularity grows you'll see more alternate DDP backends - its already happening on the front-end and the protocol is pretty straightforward.

@frozeman put in a build flag for us recently that allows you to specify a url of "default" which means just connect back in the normal meteor way (to the server that served you). When using this flag he doesn't include the Meteor.disconnect() in the index.html. Maybe this flag can also be used to control the sockjs/info issue as well.

**Comment by jeffryan** [comment]
OH... re-reading the thread I see what your saying... Your serving your app from Apache and you just want MeteorBuildClient to build a sockjs/info file for you that you can serve from your web server - along with the other client application files. 

I guess that'd make sense if the app isn't planning to ever to an explicit or implicit DDP.connect().

**Comment by dr-dimitru** [comment]
This issue is addressed in [branch `v1.0`](https://github.com/frozeman/meteor-build-client/pull/62), I encourage you to try it and report your experience.

Feel free to close it in case if the issue is solved on your end.

**Comment by dr-dimitru** [comment]
Hello @taylorbourne ,

Thank you for pointing this out!
Implemented as part of [`v1.0.0`](https://github.com/frozeman/meteor-build-client/releases/tag/1.0.0), and published last night to [NPM](http://npmjs.com/package/meteor-build-client). Please, give it a try and let me know if there's any issues.
